lib/sdf-server/src/garbage_collection.rs

```diff
@@ -0,0 +1,159 @@
+            .filter_map(|status| {
+                if status.is_active() {
+                    Some(status.to_string())
+                } else {
+                    None
+                }
+            })
+            .collect();
+        info!(
+            "Change set status(es) to consider 'active': {:?}",
+            &open_statuses
+        );
+        let change_set_snapshot_rows = ctx.txns().await?.pg().query(
+            "SELECT workspace_snapshot_address AS snapshot_id FROM change_set_pointers WHERE status = ANY($1::text[]) GROUP BY workspace_snapshot_address",
+            &[&open_statuses],
+        ).await?;
+        for row in change_set_snapshot_rows {
+            let snapshot_id: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            open_change_set_snapshot_ids.insert(snapshot_id);
+        }
+        info!(
+            "Found {} distinct snapshot address(es) for open change sets.",
+            open_change_set_snapshot_ids.len()
+        );
+
+        // Gather the WorkspaceSnapshotAddress of all existing snapshots that are
+        // at least an hour old.
+        //
+        // By only considering the snapshots that are at least an hour old, we avoid
+        // race conditions where a change set is created, or modified between when we
+        // queried the change_set_pointers table and the workspace_snapshots table.
+        // We can't rely on transactional integrity to avoid race conditions as the
+        // tables are in completely separate databases.
+        let snapshot_id_rows = ctx
+            .layer_db()
+            .workspace_snapshot()
+            .cache
+            .pg()
+            .query(
+                "SELECT key AS snapshot_id
+                 FROM workspace_snapshots
+                 WHERE created_at < NOW() - '1 hour'::interval
+                 GROUP BY key",
+                &[],
+            )
+            .await?
+            .ok_or_else(|| GarbageCollectorError::UnableToQuerySnapshots)?;
+        for row in snapshot_id_rows {
+            let snapshot_id: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            all_snapshot_ids.insert(snapshot_id);
+        }
+        info!(
+            "Found {} distinct snapshot address(es) older than cutoff.",
+            all_snapshot_ids.len()
+        );
+
+        // Any WorkspaceSnapshotAddress not in both open_change_set_snapshot_ids
+        // and all_snapshot_ids is for a closed/applied/abandoned change set and
+        // can be deleted, as we do not allow re-opening change sets.
+        let snapshot_ids_to_delete: HashSet<_> = all_snapshot_ids
+            .difference(&open_change_set_snapshot_ids)
+            .collect();
+        info!(
+            "Found {} snapshot address(es) only used by inactive change sets.",
+            snapshot_ids_to_delete.len()
+        );
+
+        let mut counter = 0;
+        for key in snapshot_ids_to_delete.iter().take(10_000) {
```

Comment:
Briefly had this as a `LIMIT 10000` in the SQL query, but realized that had the potential to limit the rows returned to only those referenced by active change sets while there might be candidates for deletion outside of the `LIMIT`. By bounding the number of items we process after having fully calculated the list, we should avoid the problem of potentially stalling out while there is work that could be done.